doc/syntax.md

```diff
@@ -2,42 +2,323 @@
+Also simple units can be put inside of square brackets.  
+
+```
+<number> <unit_type>
+<number> [<unit_type>]
+
+a mV = 1mV
+aa mV = 1[mV]
+b mV/ms = 1[mV/ms]
+```
+#### Type and unit checks
+NESTML checks type correctness of all expressions. For assignments, declarations
+with an initialization and function calls the type conformity is checked also.
+NESTML supports conversion of an `integer` to `real`. Also
+conversion between `unit`-typed and `real`-typed variables is possible. However,
+these conversions are reported as warnings. Finally, there is no conversion
+between numeric types and boolean types.
+
+### Basic elements of the embedded programming language
+Basic elements are: declarations, assignments, function calls and return statements.
+#### Declarations
+---------------------------------------------
+Declarations are composed of a non empty list of comma separated names. A
+valid name starts with a letter or underscore or dollar character. Further it
+can contain arbitrary number of letters, numbers, underscores, and dollar characters.
+Formally, a valid name satisfies the following regular expression:
+` ( 'a'..'z' | 'A'..'Z' | '_' | '$' )( 'a'..'z' | 'A'..'Z' | '_' | '0'..'9' | '$' )*`.
+Also names of functions and input ports must satisfy this pattern.
+The type of the declaration is any valid NESTML type. The type of the initialization expression
+must be compatible to the type of the declaration.
+```
+<list_of_comma_separated_names> <type> (= initialization_expression)?
+
+a, b, c real = -0.42
+d integer = 1
+n integer # default value is 0
+e string = "Bob"
+f mV   = -2e12mV
+```
+#### Assignments
+---------------------------------------------
+NESTML supports simple or compound assignments. The lefthand side of the assignments
+is always a variable. The righthand side can be an arbitrary expression of a type
+which is compatible to the lefthand side. E.g. for a numeric variable `n`:
+* simple assignment: ```n = 10```
+* compound sum: ```n += 10 corresponds to n = n + 10```
+* compound minus: ```n -= 10 corresponds to n = n - 10```
+* compound product: ```n *= 10 corresponds to n = n * 10```
+* compound quotient: ```n /= 10 corresponds to n = n - 10```
+
 
-In order to give the users of NESTML complete freedom, we provide a
-full programming language with the following features:
 
-### Physical units and data types
+#### Function calls
+---------------------------------------------
+A function call is composed of the function name and the list of required
+parameters.
 
-*  `real`, `integer`, `boolean`
-*  Physical units
-*  Type and unit checks
+```
+<function_name>(<list_of_arguments>)
+
+max(a*2, b/2)
+```
+##### Predefined Functions
+---------------------------------------------
+| | | | |
+|-|-|-| - |
+| cond_sum | curr_sum |  delta | emit\_spike |
+| exp   | expm1 | info | integrate\_odes |
+| log | max | pow | random |
+| randomInt    | resolution | steps | warning |
```

Comment:
Should we add some more comprehensive explanations here? Some of these have already been explained by @ingablundell in #4. Maybe references to the corresponding sections are enough?